### PR TITLE
feat: mark orphan runners before removing them

### PR DESCRIPTION
**Pull request by npalm**
## Problem
Orphan runners are deleted right after detection. This can be clash with self termination (ephemeral) runners. Typically the runner is waiting a few sseconds before exectuing a self termination. 

## Solution
In this solution we first mark a runner orphan, but not delete the runner. In a next cycle of the scale down function. First all orphan runners are terminated.

## Improvements

- Improved logging, only logging the main flow once at info. All other logs moved to debug
- Scale-down write permissions limitted to the envirnoment

## Todo

- [x] Update docs
- [x] Test default runner deployment
- [x] Test mult runner deployment

## Example of log

- Two instances
- One made orphan by removing the runner from GitHub
- In the log
  - Idle runner got removed
  - Orphan get marked as orphan
  - Next cycle orphan terminated.  
image

### CODE REVIEW

File: lambdas/functions/control-plane/src/aws/runners.ts [hunk 1]
```diff
@@ -46,6 +48,9 @@ function constructFilters(filters?: Runners.ListRunnerFilters): Ec2Filter[][] {
       ec2FiltersBase.push({ Name: `tag:ghr:Type`, Values: [filters.runnerType] });
       ec2FiltersBase.push({ Name: `tag:ghr:Owner`, Values: [filters.runnerOwner] });
     }
+    if (filters.orphan) {
+      ec2FiltersBase.push({ Name: 'tag:ghr:orphan', Values: ['true'] });
```
Comment (stuartp44) [inline lambdas/functions/control-plane/src/aws/runners.ts:52]:
```suggestion
      ec2FiltersBase.push({ Name: 'tag:ghr:Orphan', Values: ['true'] });
```
Everything else uses caps?

File: lambdas/functions/control-plane/src/aws/runners.ts [hunk 2]
```diff
@@ -85,6 +90,7 @@ function getRunnerInfo(runningInstances: DescribeInstancesResult) {
             type: i.Tags?.find((e) => e.Key === 'ghr:Type')?.Value as string,
             repo: i.Tags?.find((e) => e.Key === 'ghr:Repo')?.Value as string,
             org: i.Tags?.find((e) => e.Key === 'ghr:Org')?.Value as string,
+            orphan: i.Tags?.find((e) => e.Key === 'ghr:orphan')?.Value === 'true',
```
Comment (stuartp44) [inline lambdas/functions/control-plane/src/aws/runners.ts:93]:
```suggestion
            orphan: i.Tags?.find((e) => e.Key === 'ghr:Orphan')?.Value === 'true',
```
As Above

File: lambdas/functions/control-plane/src/scale-runners/scale-down.test.ts [hunk 3]
```diff
@@ -293,10 +295,31 @@ describe('Scale down runners', () => {
...
+        // act
+        await scaleDown();
+
+        // assert
+        checkTerminated(runners);
+        checkNonTerminated(runners);
+
+        expect(mockTagRunners).toHaveBeenCalledWith(orphanRunner.instanceId, [
+          {
+            Key: 'orphan',
```
Comment (stuartp44) [inline lambdas/functions/control-plane/src/scale-runners/scale-down.test.ts:314]:
```suggestion
            Key: 'Orphan',
```
As above?

File: lambdas/functions/control-plane/src/scale-runners/scale-down.ts [hunk 4]
```diff
@@ -194,11 +193,26 @@ async function evaluateAndRemoveRunners(
   }
 }
 
-async function terminateOrphan(instanceId: string): Promise<void> {
+async function markOrphan(instanceId: string): Promise<void> {
   try {
-    await terminateRunner(instanceId);
+    await tag(instanceId, [{ Key: 'orphan', Value: 'true' }]);
```
Comment (stuartp44) [inline lambdas/functions/control-plane/src/scale-runners/scale-down.ts:198]:
```suggestion
    await tag(instanceId, [{ Key: 'Orphan', Value: 'true' }]);
```
as above

File: lambdas/functions/control-plane/src/scale-runners/scale-down.ts [hunk 5]
```diff
@@ -194,11 +193,26 @@ async function evaluateAndRemoveRunners(
...
+  }
+}
+
+async function terminateOrphan(environment: string): Promise<void> {
+  try {
+    const orphanRunners = await listEC2Runners({ environment, orphan: true });
+
+    for (const runner of orphanRunners) {
+      await terminateRunner(runner.instanceId).catch((e) => {
+        logger.error(`Failed to terminate orphan runner '${runner.instanceId}'`, { error: e });
```
Comment (stuartp44) [inline lambdas/functions/control-plane/src/scale-runners/scale-down.ts:211]: Is there any need to do a catch twice?

File: lambdas/functions/control-plane/src/scale-runners/scale-down.ts [hunk 6]
```diff
@@ -221,14 +235,18 @@ async function listRunners(environment: string) {
...
 function filterRunners(ec2runners: RunnerList[]): RunnerInfo[] {
-  return ec2runners.filter((ec2Runner) => ec2Runner.type) as RunnerInfo[];
+  return ec2runners.filter((ec2Runner) => ec2Runner.type && !ec2Runner.orphan) as RunnerInfo[];
 }
 
 export async function scaleDown(): Promise<void> {
   githubCache.reset();
-  const scaleDownConfigs = JSON.parse(process.env.SCALE_DOWN_CONFIG) as [ScalingDownConfig];
   const environment = process.env.ENVIRONMENT;
+  const scaleDownConfigs = JSON.parse(process.env.SCALE_DOWN_CONFIG) as [ScalingDownConfig];
```
Comment (stuartp44) [inline lambdas/functions/control-plane/src/scale-runners/scale-down.ts:244]: I have noted here that we parse without sanity checking it first?

File: lambdas/functions/control-plane/src/scale-runners/scale-down.ts (hunk 5 above)
Comment (stuartp44) [inline lambdas/functions/control-plane/src/scale-runners/scale-down.ts:211]: Is there any point doing a catch twice? maybe worth doing it once in the catch block

**Review by stuartp44 — CHANGES_REQUESTED** [review]

File: lambdas/functions/control-plane/src/scale-runners/scale-down.ts (hunk 6 above)
Comment (npalm) [reply lambdas/functions/control-plane/src/scale-runners/scale-down.ts:244]: The config is typed in Terraform, but it is a loose coupling.

File: lambdas/functions/control-plane/src/scale-runners/scale-down.ts (hunk 5 above)
Comment (npalm) [reply lambdas/functions/control-plane/src/scale-runners/scale-down.ts:211]: Did quite some testing with it, but it needed due to all the async executing.

File: lambdas/functions/control-plane/src/aws/runners.ts (hunk 1 above)
Comment (npalm) [reply lambdas/functions/control-plane/src/aws/runners.ts:52]: Most of the tags we use are in lower case like `ghr:environment`, `ghr:ssm_config_path` would suggest we stick to lower case an update owner and type to lower as well in a sperate PR.

File: lambdas/functions/control-plane/src/scale-runners/scale-down.ts (hunk 4 above)
Comment (npalm) [reply lambdas/functions/control-plane/src/scale-runners/scale-down.ts:198]: in fact this should be `ghr:orphan`

File: lambdas/functions/control-plane/src/scale-runners/scale-down.ts (hunk 5 above)
Comment (npalm) [reply lambdas/functions/control-plane/src/scale-runners/scale-down.ts:209]: yes the inner try/catch is catchting error during termination. The outer any other inlcuding the listEC2Runners

File: lambdas/functions/control-plane/src/scale-runners/scale-down.ts (hunk 4 above)
Comment (npalm) [reply lambdas/functions/control-plane/src/scale-runners/scale-down.ts]: created #4026 for formatting of the keuys.

**Review by Brend-Smits — APPROVED** [review]
LGTM!

**Pull request merged**